include/pybind11/stl.h

```diff
@@ -166,13 +166,38 @@ struct list_caster {
     using value_conv = make_caster<Value>;
 
     bool load(handle src, bool convert) {
-        if (!isinstance<sequence>(src) || isinstance<bytes>(src) || isinstance<str>(src)) {
+        if (isinstance<bytes>(src) || isinstance<str>(src)) {
             return false;
         }
-        auto s = reinterpret_borrow<sequence>(src);
+        if (isinstance<sequence>(src)) {
+            return convert_elements(src, convert);
+        }
+        if (!convert) {
+            return false;
+        }
+        if (PyGen_Check(src.ptr())) {
```

Comment:
Shouldn't we really be supporting a `py::iterable` object instead? I think a generator would count here and simplify the code.

Reply:
I definitely like the general thinking.

Please give me a little more time. I'm in the middle of what I am thinking of as "data driven cleanup".

PyCLIF is very far on the permissive side of the spectrum. — That's what's giving me the data: what is being used, how much.

pybind11 is very far on the strict side.

A glimpse on where I am right now: PyCLIF supports passing any iterable for a C++ set. I think that's going too far to the permissive side of the spectrum. So I'm going through cleaning that up, even though it's a lot of leg work. But I believe it makes the code safer and more readable.

Reply:
I just pushed the equivalent of the behavior I settled on for PyCLIF. I'm still working on rolling out the PyCLIF behavior changes internally. Only after that can I submit internally and then export the changed PyCLIF to github.

High-level: I'm converging the behaviors

* by changing PyCLIF to be stricter & safer,
* while this PR makes pybind11 more user friendly without crossing into unsafe territory.

> Shouldn't we really be supporting a `py::iterable` object instead?

That's what PyCLIF used to do, but it has pitfalls that led to lots of unclean test code: Python `{}` (empty dict) passed as input for `std::vector` and `std::set`, `[]` as input for `std::set` and `std::map`.

The originally very permissive PyCLIF behavior gave me a fairly unique set of data: I could see what's being used a lot, what makes sense (IMO), and what doesn't make sense or seems unsafe (again IMO). The new code block near the top of stl.h, transferred here from PyCLIF, encodes my conclusions. With that the internal cleanup turned out to be relatively small, only around 50 changes I had to mail out. This is part of the change description for those (slightly edited):

________

Motivations for the behavior change:

* Improves safety / prevents accidents:

  * For C++ `std::set` (and `std::unordered_set`): The potentially reducing conversion from a Python sequence (e.g. Python `list` or `tuple`) to a C++ set must be explicit.

  * For C++ `std::vector`: In very rare cases (`itertools.chain`) the conversion from a Python iterable must be explicit.

* Improves readability: Python literals to be passed as C++ sets must be set literals.
________


I'll finish the rollout of the PyCLIF change, then come back here for polishing maybe and further expanding the tests. Please let me know any high-level suggestions.

Reply:
The PyCLIF behavior change (with prior cleanup around the codebase) is now complete: https://github.com/google/clif/commit/30c54882eaa884a502384e1ec3a4e2f1e9dd85e1

Internally I'm testing (globally) PyCLIF-pybind11 with this PR. Unfortunately some unrelated problem slipped in that I need to troubleshoot first, before I can convince myself that this PR actually fixes all pybind11/stl.h vs PyCLIF behavior differences.